### ISSUE TITLE
memory leak in dsl_scan_visitbp()

### ISSUE THREAD

**Issue by rottegift** [openzfsonosx/zfs#602]
14bf64a6e04 introduced this change:

```
@@ -814,8 +903,8 @@ dsl_scan_visitbp(blkptr_t *bp, const zbookmark_phys_t *zb,

        /* ASSERT(pbuf == NULL || arc_released(pbuf)); */

-       if (dsl_scan_check_pause(scn, zb))
-               goto out;
+       if (dsl_scan_check_suspend(scn, zb))
+               return;
```

But note that this is after the kmem_alloc(), and below out: is the corresponding kmem_free.

The result is blkptr_t leaks during scans (scrubs, resilvers) that fill up the kmem_alloc_128 kmem_cache.

**Comment by rottegift** [comment]
Openzfs's dsl_scan_visitbp doesn't do any allocating.

Zol's dsl_scan_visitbp does:

```
        if (dsl_scan_check_suspend(scn, zb))
                goto out;
```

So this is a porting from openzfs to openzfsonosx problem.  :-(

**Comment by rottegift** [comment]
For posterity:

Discovered the leaker via poor performance.   sysctl kstat.spl.misc.spl_misc.os_mem_alloc was extremely high compared to the ARC.   Ran a sort on sysctl kstat.unix.kmem_cache.buf_inuse and found that kmem_alloc_128 was many GB (it should be no more than a few MB in ordinary circumstances).

In the testing setup, there would be a small but bounded leak, with the leak ending after the tester stopped.

The kmem_alloc_128 cache holds objects from kmem_alloc(size) (and kmem_zalloc(size)) where size is between 65 and 128 bytes inclusive.   Unfortunately there are many objects that are allocated in that size range, and object lifetimes can be spread across vastly different functions (e.g. create at one end of a zio pipeline with destroy at the other end, or at a callback).

After some thought, I opened three windows and ran concurrently:

```
sysctl kstat.unix.kmem_cache.buf_inuse
```
and
```
sudo dtrace -n 'zfs_kmem_*alloc:entry /arg0 <= 128 && arg0 > 64/ {  @[caller] = count(); } END { printa("%40a %@16d\n", @); }'
```
and
```
sudo dtrace -n 'zfs_kmem_free:entry /arg1 <= 128 && arg1 > 64/ {  @[caller] = count(); } END { printa("%40a %@16d\n", @); }'
```

for about two minutes.   In dsl_scan_visitbp there were about as many missing frees as the increase in the kmem_alloc_128 kmem_cache.  

I was running a zpool scrub at the time.   Pausing the scrub made the kmem_alloc_128 cache stop growing.

The cause of the leak was obvious once it was tracked down to dsl_scan_visitbp.
